### PR TITLE
feat: add minimumReleaseAge supply-chain security check

### PR DESCRIPTION
**Pull request by huhuanming**
## Summary

- Add a `minimumReleaseAge` check that blocks npm packages published less than N days ago (default 7), similar to Bun's `minimumReleaseAge`
- Diff-based approach: only checks new/changed packages in `yarn.lock` via `git diff`, not the full lockfile (~10k entries), so typical PR checks complete in 1-2s
- Supports glob-pattern allowlist (`@onekeyhq/*`, `@onekeyfe/*`), configurable threshold and block/warn mode via `package.json`
- Includes `full` scan mode for periodic audits of the entire lockfile
- 33 unit tests covering core logic, diff parsing, edge cases

## Usage

```bash
# Diff mode (fast, for CI) — checks only new packages vs base branch
yarn check:release-age

# Full audit mode — checks all packages in yarn.lock
yarn check:release-age:full
```

## Configuration (package.json)

```json
"minimumReleaseAge": {
  "days": 7,
  "allowlist": ["@onekeyhq/*", "@onekeyfe/*"],
  "blockOnFailure": true,
  "registryUrl": "https://registry.npmjs.org"
}
```

## Test plan

- [x] All 33 unit tests pass (`npx jest development/scripts/minimum-release-age/__tests__/`)
- [x] Smoke test `yarn check:release-age` on a branch with lockfile changes
- [ ] Verify CI integration when added to lint workflow

---

<a href="https://app.devin.ai/review/onekeyhq/app-monorepo/pull/11010" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by revan-zhang** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/a51de69a-463b-44ce-ad38-d043d70d8b2c?refs=3ffa340d1ea53dbb7277dc56437735d71914a33b&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/a51de69a-463b-44ce-ad38-d043d70d8b2c/license?refs=3ffa340d1ea53dbb7277dc56437735d71914a33b&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: development/scripts/minimum-release-age/run.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+// --- Constants ---
+
+const BATCH_SIZE = 10;
+const DEFAULT_BASE_BRANCH = 'x';
+
+// --- Registry fetch ---
+
+function fetchRegistryMeta(name: string, registryUrl: string): Promise<NpmPackageMeta> {
+  const url = `${registryUrl}/${encodeURIComponent(name).replace('%40', '@')}`;
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply development/scripts/minimum-release-age/run.ts]: Fixed in e4d0562 — changed `replace` to `replaceAll` to ensure all occurrences of `%40` are replaced.

File: .github/workflows/security-checks.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Install Dependency
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NODE_OPTIONS: '--max_old_space_size=4096'
+        run: |
+          yarn install --immutable
+
+      - name: Check minimum release age
+        run: |
+          npx tsx development/scripts/minimum-release-age/run.ts diff origin/${{ github.base_ref }}
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply .github/workflows/security-checks.yml:60]: Fixed in e4d0562 — added `permissions: contents: read` to the workflow.

File: development/scripts/minimum-release-age/parseLockfileDiff.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * the package name and version, then deduplicate.
+ */
+export function parseLockfileDiff(diff: string): PackageRef[] {
+  // Match added lines containing a package entry.
+  // Captures: (optional @scope/name or plain name) and (version after npm:)
+  // Examples:
+  //   +"lodash@npm:4.17.21":
+  //   +"@babel/core@npm:7.24.0":
+  //   +"@scope/name@npm:1.2.3, @scope/name@npm:^1.0.0":
+  const pattern = /^\+"(?:(@[^@]+\/[^@]+)|([^@"][^@]*))@npm:(\d+\.\d+[^",]*)"?/gm;
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply development/scripts/minimum-release-age/parseLockfileDiff.ts]: Fixed in 3a902b3 — rewrote `parseLockfileDiff` to use a two-step approach (match entry header for package name, then read resolved version from the `+  version:` line), same strategy as `parseFullLockfile`. Now correctly handles `^`, `~`, `>=` range specifiers in entry headers.

File: development/scripts/minimum-release-age/run.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  if (mode === 'full') {
+    console.log('Running full lockfile audit...');
+    const lockfilePath = path.join(rootDir, 'yarn.lock');
+    const lockfileContent = fs.readFileSync(lockfilePath, 'utf-8');
+    packages = parseFullLockfile(lockfileContent);
+  } else {
+    console.log(`Running diff check against ${baseBranch}...`);
+    let diff: string;
+    try {
+      diff = execSync(`git diff ${baseBranch} -- yarn.lock`, {
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply development/scripts/minimum-release-age/run.ts]: Fixed in 3a902b3 — replaced `execSync` with `execFileSync('git', ['diff', baseBranch, '--', 'yarn.lock'])` to avoid shell interpretation of the `baseBranch` argument.

File: development/scripts/minimum-release-age/run.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    try {
+      diff = execFileSync('git', ['diff', baseBranch, '--', 'yarn.lock'], {
+        cwd: rootDir,
+        encoding: 'utf-8',
+        maxBuffer: 50 * 1024 * 1024,
+      });
+    } catch {
+      console.log('No yarn.lock changes detected or git diff failed.');
+      process.exit(0);
+    }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply development/scripts/minimum-release-age/run.ts]: Fixed in 80d57f5 — the catch block now calls `process.exit(1)` instead of `process.exit(0)`, so the security gate fails closed when git diff encounters an error.

File: development/scripts/minimum-release-age/run.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+  const results = await processBatches(packages, config);
+  const hasTooYoung = printResults(results, config);
+
+  if (hasTooYoung && config.blockOnFailure) {
+    console.log(
+      'Blocking: one or more packages are younger than the minimum release age threshold.',
+    );
+    process.exit(1);
+  }
```
[the remark this answers is not in the document]
Comment (huhuanming) [reply development/scripts/minimum-release-age/run.ts]: Fixed in 6af7d60 — `printResults` now returns both `hasTooYoung` and `hasErrors`, and the main function calls `process.exit(1)` when either is true. Registry fetch errors now fail closed, consistent with the git diff error handling.

**Review by sidmorizon — APPROVED** [review]

**Review by limichange — APPROVED** [review]
